src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -30,10 +30,27 @@ case class MatrixType(
 
   def genotypeType: Type = metadata.genotypeSignature
 
-  def rowType: TStruct = TStruct(
-    "v" -> vType,
-    "va" -> vaType,
-    "gs" -> TArray(genotypeType))
+  def partitionKeyType: Type =
+    vType match {
+      case t: TVariant =>
+        TLocus(vType.asInstanceOf[TVariant].gr)
+      case _ =>
+        vType
+    }
```

Comment:
This pattern also appears in VariantSampleMatrix. I'm not sure there's a natural way to avoid this duplication...